keps/1714-fair-sharing/README.md

```diff
@@ -0,0 +1,366 @@
+their siblings, and should get proportionally more resources than others located
+under the same parent. 
+
+* When distributing resources, prioritize needs of CQ within the same
+sub-organization and only after they have been fulfilled, proceed with distribution
+outside of the suborganization.
+
+* When enforcing fair sharing, ignore workload priorities unless:
+
+** The workload's priority is above admin-defined high priority. Super high
+priority workloads overrule fair sharing and are treated according to KEP #1337.
+
+** There is a need to preempt some non top priority workload from a ClusterQueue.
+Then the lowest priority workloads from a CQ that is over its fair share should
+be selected for preemption.
+
+* The mechanism should be compatible with all existing or proposed Kueue features,
+in particular:
+
+** Borrowing limits 
+** Lending limits
+** Guaranteed/nominal quota
+** Hierarchical cohorts
+
+* Fair sharing should not limit Kueue scalability. Kueue, with fair sharing enabled,
+should be able to handle >1k ClusterQueues, >100 Cohorts and >10k workloads (that
+are either running or queued) within a single hierarchical organization.
+
+* The proposed system should be hard to game, for example by creating big workloads
+that consume all capacity.
+
+* Fair sharing enforcement should not significantly decrease overall utilization,
+however, pathological situations (like a single workload consuming all otherwise
+unused capacity) should be resolved in favor of fair sharing than maximizing the
+utilization (the big greedy workload should be preempted to admit smaller
+workloads from other CQ that consume only their fair share).
+
+### Non-Goals
+* Use historical data (for example CQ A used a lot of shared capacity for the last
+week, so now it should get less because others, who didn't need anything then, have
+pending workloads). Fair sharing should be based on point-in-time situation, although,
+ideally it should be expandable to support history-based fair sharing without major redesign.
+
+* Enable fair sharing only for some part of the resources or Cohort hierarchy.
+Fair sharing will be a global switch (at least initially).
+
+* Maximize utilization at the cost of fair sharing.
+
+## Proposal
+Introduce a global fair sharing mechanism that is based on preemptions. As long
+as there are some free and accessible resources in the cohort hierarchy, Kueue
+will admit workloads without any limits. However, once the capacity is gone,
+new workloads from Cohorts/CQ that have not received their fair share yet will
+be permitted to preempt workloads from Chorots/CQ that received more than their
+fair share would otherwise be (or more than the Cohort/CQ that is in need).
+
+Additionally, we will introduce a reclaim mechanism at the Cohort level. Cohorts
+that are in need, but whose consumption didn’t reach their total aggregated
+nominal quota will be able to preempt workloads from Cohorts/CQ that are over
```

Comment:
What about ClusterQueues which reached the total aggregated quota, aren't they able to preempt? This scenario was one of the motivations for https://github.com/kubernetes-sigs/kueue/tree/main/keps/1337-preempt-within-cohort-while-borrowing, in which nominalQuota for ClusterQueues is 0 (so already reached), yet they are able to preempt. I'm wondering if we are not losing this scenario here.